topi/python/topi/x86/tensor_intrin.py

```diff
@@ -275,3 +275,97 @@ def _instr(index):
+        The AVX2 int8 TensorIntrin that can be used in tensorizing schedule
+    """
+
+    int32_lanes = 16 # 16 int32 lanes in AVX2
+    num_int8_elements = 4 # 4 int8 elements in int32
+    data = tvm.placeholder((num_int8_elements,), dtype='uint8', name='data')
+    kernel = tvm.placeholder((int32_lanes, num_int8_elements), dtype='int8', name='kernel')
+    k = tvm.reduce_axis((0, num_int8_elements), name='k')
+    C = tvm.compute((int32_lanes,),
+                    lambda i: tvm.sum(data[k].astype('int32') *
+                                      kernel[i, k].astype('int32'),
+                                      axis=k),
+                    name="C")
+
+    a_buffer = tvm.decl_buffer(data.shape, dtype='uint8', name="a_buffer",
+                               offset_factor=1,
+                               strides=[1])
+    b_buffer = tvm.decl_buffer(kernel.shape, dtype='int8', name="b_buffer",
+                               offset_factor=1,
+                               strides=[tvm.var('ldw'), 1])
+
+    def _intrin_func(ins, outs):
+        def _instr(index):
+            ib = tvm.ir_builder.create()
+            if index == 1:
+                ib.emit(outs[0].vstore(0, tvm.const(0, 'int32x16')))
+                return ib.get()
+
+            a_int8 = ins[0].vload([0], "uint8x4")
+            re_int32 = tvm.call_pure_intrin('int32', 'reinterpret', a_int8)
+            vec_ai32 = re_int32.astype('int32x8')
+            vec_a = tvm.call_pure_intrin('int8x32', 'reinterpret', vec_ai32)
+            vec_b_0 = ins[1].vload([0, 0], "int8x32")
+            vec_b_1 = ins[1].vload([8, 0], "int8x32")
+            vec_one = tvm.const(1, "int16x16")
+            pair_reduction_0 = tvm.call_llvm_intrin('int16x16',
+                                                  'llvm.x86.avx2.pmadd.ub.sw',
+                                                  tvm.const(0, 'uint32'),
+                                                  vec_a, vec_b_0)
+            quad_reduction_0 = tvm.call_llvm_intrin('int32x8',
+                                                  'llvm.x86.avx2.pmadd.wd',
+                                                  tvm.const(0, 'uint32'),
+                                                  pair_reduction_0, vec_one)
+            pair_reduction_1 = tvm.call_llvm_intrin('int16x16',
+                                                  'llvm.x86.avx2.pmadd.ub.sw',
+                                                  tvm.const(0, 'uint32'),
+                                                  vec_a, vec_b_1)
+            quad_reduction_1 = tvm.call_llvm_intrin('int32x8',
+                                                  'llvm.x86.avx2.pmadd.wd',
+                                                  tvm.const(0, 'uint32'),
+                                                  pair_reduction_1, vec_one)
+            if index == 0:
+                ib.emit(outs[0].vstore([0], quad_reduction_0))
+                ib.emit(outs[0].vstore([8], quad_reduction_1))
```

Comment:
Why have 2 quad_reductions in assembly instead of unrolling in the schedule?

Reply:
Oh maybe, you would have then change num_lanes etc

If thats tha case, @yzhliu is working on making it an argument. Maybe we can use that to simplify this.

Reply:
@anijain2305. Correct. I wanted to be able to use the same memory layout which can be tensorized for avx2 and avx512.